osquery/events/windows/etw/etw_kernel_session.cpp

```diff
@@ -0,0 +1,210 @@
+/**
+ * Copyright (c) 2014-present, The osquery authors
+ *
+ * This source code is licensed as defined by the LICENSE file found in the
+ * root directory of this source tree.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR GPL-2.0-only)
+ */
+
+#include <osquery/core/flags.h>
+#include <osquery/events/windows/etw/etw_kernel_session.h>
+#include <osquery/events/windows/etw/etw_provider_config.h>
+#include <osquery/utils/conversions/windows/strings.h>
+#include <osquery/utils/status/status.h>
+
+namespace osquery {
+
+HIDDEN_FLAG(
+    uint32,
+    etw_kernel_trace_buffer_size,
+    kEventTraceBufferSize,
+    "Kilobytes of memory allocated for the ETW kernelspace tracing session");
+
+HIDDEN_FLAG(
+    uint32,
+    etw_kernel_trace_minimum_buffers,
+    kEventTraceMinimumBuffers,
+    "Minimum number of buffers reserved for the tracing session buffer pool");
+
+HIDDEN_FLAG(
+    uint32,
+    etw_kernel_trace_maximum_buffers,
+    kEventTraceMaximumBuffers,
+    "Maximum number of buffers reserved for the tracing session buffer pool");
+
+HIDDEN_FLAG(uint32,
+            etw_kernel_trace_flush_timer,
+            kEventTraceFlushTimer,
+            "How often, in seconds, any non-empty trace buffers are flushed");
+
+KernelEtwSessionRunnable::KernelEtwSessionRunnable(
+    const std::string& runnableName)
+    : InternalRunnable(runnableName) {
+  initKernelTraceSession(runnableName);
+}
+
+KernelEtwSessionRunnable::~KernelEtwSessionRunnable() {
+  stop();
+
+  if (kernelTraceSession_) {
+    kernelTraceSession_.reset();
+  }
+}
+
+Status KernelEtwSessionRunnable::addProvider(
+    const EtwProviderConfig& configData) {
+  // Sanity check on input ETW Provider configuration data
+  Status validProvider = configData.isValid();
+  if (!validProvider.ok()) {
+    return Status::failure("Invalid ETW provider configuration data: " +
+                           validProvider.getMessage());
+  }
+
+  // Sanity check on trace session object
+  if (!kernelTraceSession_) {
+    return Status::failure("ETW kernel trace session is not initialized");
+  }
+
+  KernelProviderRef kernelProvider{nullptr};
+
+  // Supported kernel providers instantiation
+  EtwProviderConfig::EtwKernelProviderType kProviderType =
+      configData.getKernelProviderType();
+
+  switch (kProviderType) {
+  case EtwProviderConfig::EtwKernelProviderType::File: {
+    // https://learn.microsoft.com/en-us/windows/win32/etw/fileio
+    kernelProvider = std::make_shared<krabs::kernel::file_io_provider>();
+  } break;
```

Comment:
Nit: I think you want these breaks inside of the curlys, keep all the code for each case together inside.

Reply:
good point; I've just fixed that.